sqlmesh/integrations/github/cicd/controller.py

```diff
@@ -0,0 +1,717 @@
+        )
+
+    @property
+    def formatted_loaded_intervals(self) -> str:
+        merged_inclusive_intervals = [
+            make_inclusive(start, end) for start, end in merge_intervals(self.intervals)
+        ]
+        return ", ".join(
+            f"({_format_date_time(start, self.interval_unit)} - {_format_date_time(end, self.interval_unit)})"
+            for start, end in merged_inclusive_intervals
+        )
+
+    @property
+    def change_category_str(self) -> str:
+        return SNAPSHOT_CHANGE_CATEGORY_STR[self.change_category]
+
+
+class GithubEvent:
+    """
+    Takes a Github Actions event payload and provides a simple interface to access
+    """
+
+    def __init__(self, payload: t.Dict[str, t.Any]) -> None:
+        self.payload = payload
+        self._pull_request_info: t.Optional[PullRequestInfo] = None
+
+    @classmethod
+    def from_path(cls, path: t.Union[str, pathlib.Path]) -> GithubEvent:
+        with open(path) as f:
+            return cls(payload=json.load(f))
+
+    @classmethod
+    def from_env(cls) -> GithubEvent:
+        return cls.from_path(os.environ["GITHUB_EVENT_PATH"])
+
+    @property
+    def is_review(self) -> bool:
+        return bool(self.payload.get("review"))
+
+    @property
+    def is_comment(self) -> bool:
+        return bool(self.payload.get("comment"))
+
+    @property
+    def is_pull_request(self) -> bool:
+        return bool(self.payload.get("pull_request"))
+
+    @property
+    def is_pull_request_closed(self) -> bool:
+        return self.is_pull_request and self.payload.get("action") == "closed"
+
+    @property
+    def pull_request_url(self) -> str:
+        if self.is_review:
+            return self.payload["review"]["pull_request_url"]
+        if self.is_comment:
+            return self.payload["issue"]["pull_request"]["url"]
+        if self.is_pull_request:
+            return self.payload["pull_request"]["_links"]["self"]["href"]
+        raise CICDBotError("Unable to determine pull request url")
+
+    @property
+    def pull_request_info(self) -> PullRequestInfo:
+        if not self._pull_request_info:
+            self._pull_request_info = PullRequestInfo.create_from_pull_request_url(
+                self.pull_request_url
+            )
+        return self._pull_request_info
+
+
+class GithubController:
+    def __init__(
+        self,
+        paths: t.Union[str, t.Iterable[str]],
+        token: str,
+        config: t.Optional[t.Union[Config, str]] = None,
+        event: t.Optional[GithubEvent] = None,
+    ) -> None:
+        self._paths = paths
+        self._config = config
+        self._token = token
+        self._event = event or GithubEvent.from_env()
+        self._pr_plan: t.Optional[Plan] = None
+        self._prod_plan: t.Optional[Plan] = None
+        self._console: t.Optional[MarkdownConsole] = None
+        self._check_run_mapping: t.Dict[str, CheckRun] = {}
+        self.__client: t.Optional[Github] = None
```

Comment:
what's the point of the double underscore here vs single?

Reply:
ah, you're caching

Reply:
it doesn't seem like there's a point to caching these, with the properties, it's just extra code, just initialize them inline here

does it really need to be lazy?

Reply:
I don't see a negative with the lazy approach. Isolate all of the logic to each property and then only execute if needed.

Reply:
it makes the code way more bloated half the code is this boiler plate, i'd much prefer it if we kept things simpler